### ISSUE TITLE
CL setting in calculating training loss in v0.2.0 is wrong

### ISSUE THREAD

**Issue by wanjunhong0** [zezhishao/BasicTS#60]
It used to be right. But after v0.2.0 change. 
In base_tsf_runner.py #292 line the metric to calcuate loss should the forward_return[:2] which affected by CL setting. 
But now it does not. And the model performance is worse than before because of this. 
NEED TO FIX ASAP.

**Comment by zezhishao** [comment]
Thanks for your advice. Does it really affect the ***final*** perfomance? 
The line #292 should not affect the ***final*** performance, since we only calculate loss based on current curriculum learning length. See line #288. 
Chaning the line #292 aims to get consistent logs for both training/valid/test datasets.

**Comment by wanjunhong0** [comment]
But the the code from #281~286 has really no effect. Why does the CL flag in training anyway? The CL setting is only effective on validing and testing now. Is CL should be working in the training process? Now it is only masking the testing results.
And the loss from #288 is different from #291~292...there is something wrong here.

**Comment by zezhishao** [comment]
I think there might be some misunderstandings. Line 288 uses the `forward_return` to calculate loss, which is used to conduct back propagation. Thus, the CL does have effect. Line 291 and 292 are used to calculate the metrics, which are used for evaluating the performance and **not** for conducting back propagation.

**Comment by wanjunhong0** [comment]
OK. I can see that. 

But there must be something wrong here. 
I run your STEP code by upgrading the basicTS to the newest version and the performance was worse about 0.5 on MAE (that's a lot!) on all horizons. But it was fine with the basicTS that STEP version updated five months ago.
And that's how I find the problem. I notice the train MAE was significantly larger. Now the issue is not here. 

What else changed from v0.2.0 to the newest version?

**Comment by zezhishao** [comment]
> OK. I can see that.
> 
> But there must be something wrong here. I run your STEP code by upgrading the basicTS to the newest version and the performance was worse about 0.5 on MAE (that's a lot!) on all horizons. But it was fine with the basicTS that STEP version updated five months ago. And that's how I find the problem. I notice the train MAE was significantly larger. Now the issue is not here.
> 
> What else changed from v0.2.0 to the newest version?

Which datasets?

**Comment by zezhishao** [comment]
> OK. I can see that.
> 
> But there must be something wrong here. I run your STEP code by upgrading the basicTS to the newest version and the performance was worse about 0.5 on MAE (that's a lot!) on all horizons. But it was fine with the basicTS that STEP version updated five months ago. And that's how I find the problem. I notice the train MAE was significantly larger. Now the issue is not here.
> 
> What else changed from v0.2.0 to the newest version?

Which dataset?

**Comment by wanjunhong0** [comment]
> > OK. I can see that.
> > But there must be something wrong here. I run your STEP code by upgrading the basicTS to the newest version and the performance was worse about 0.5 on MAE (that's a lot!) on all horizons. But it was fine with the basicTS that STEP version updated five months ago. And that's how I find the problem. I notice the train MAE was significantly larger. Now the issue is not here.
> > What else changed from v0.2.0 to the newest version?
> 
> Which dataset?

METR-LA

**Comment by zezhishao** [comment]
I'm so sorry for these bugs, because I don't have much more time to add detailed documentation and changelogs for BasicTS. 
I'm guessing the most likely cause is the `CFG.TRAIN.NULL_VAL` option.
In later versions of BasicTS, `CFG.TRAIN.NULL_VAL` was changed to `CFG.NULL_VAL`. 
This is because the `null_val` in metrics (such as [masked_mae](https://github.com/zezhishao/BasicTS/blob/d2059631c7353da56960ac97985b5187c1cb26b1/basicts/losses/losses.py#L22)) is used across all processes, not just TRAIN.

I would like to strongly recommend you to use the BasicTS from the repo in STEP, so as not to cause more bugs. This is also why I update the BasicTS folder locally instead of installing via pip.

**Comment by zezhishao** [comment]
Please feel free to ask any other questions, I will reply as soon as possible.

**Comment by wanjunhong0** [comment]
Thanks. I can confirm that adding CFG.NULL_VAL = 0. fixed my problem. 
And CL in current version is fine. 

I am closing this issue.
